src/rulesets/Base/array.jl

```diff
@@ -81,10 +81,11 @@ function rrule(::typeof(hcat), Xs::Union{AbstractArray, Number}...)
     Y = hcat(Xs...)  # note that Y always has 1-based indexing, even if X isa OffsetArray
     ndimsY = Val(ndims(Y))  # this avoids closing over Y, Val() is essential for type-stability
     sizes = map(size, Xs)   # this avoids closing over Xs
+    project_Xs = map(ProjectTo, Xs)
     function hcat_pullback(ȳ)
         dY = unthunk(ȳ)
         hi = Ref(0)  # Ref avoids hi::Core.Box
-        dXs = map(sizes) do sizeX
+        dXs = map(project_Xs, sizes) do project, sizeX
             ndimsX = length(sizeX)
```

Comment:
Seems like a lot of boilerplate to add this to every rule definition by hand. Can't the computer do this?

Reply:
Maybe, but it's beyond the scope of this PR.
This is the much reduced boiler plate version.

Reply:
Possibly we could, though there are a few complications.
1) project needs to be inside thunks, because projecting the thunks would unthunk them (unless we want to nest thunks, which I am told can be a bad thing "because compiler")
2) we'd have to check whether the performance is worse for rules that don't actually need projecting

Reply:
What I mean is that this is a high-level design question, whether this thing is to be globally applied, or is something to be selectively hand-added to rules which need it. I don't know if I missed this, I couldn't follow everything last week, but was this discussed or did it just sort-of happen?

There was a working implementation here, which I think is applied after un-thunking:
https://github.com/FluxML/Zygote.jl/pull/965/files#diff-e0bc7da8f1a33a59f5ecfa67257c04038f0b4915b3f74bdf39780818fd0010a2R73-R97